src/software/power/geneva.cpp

```diff
@@ -1,32 +1,128 @@
+void (*volatile IRAM_ATTR Geneva::rotation_done_callback)() = NULL;
 
 Geneva::Geneva()
 {
-    timer = timerBegin(0, 80, true);
+    pinMode(DIR, OUTPUT);
+    pinMode(PWM, OUTPUT);
+    pinMode(GENEVA_ENCODER_A_PIN, INPUT_PULLDOWN);
+    pinMode(GENEVA_ENCODER_B_PIN, INPUT_PULLDOWN);
+
+    attachInterrupt(digitalPinToInterrupt(GENEVA_ENCODER_A_PIN), pulseEncoderA, RISING);
+    attachInterrupt(digitalPinToInterrupt(GENEVA_ENCODER_B_PIN), pulseEncoderB, RISING);
+
+    timer = timerBegin(GENEVA_TIMER, 80, true);
+
+    timerAttachInterrupt(timer, &onTimer, true);
+    // call onTimer every 0.1 seconds
+    timerAlarmWrite(timer, 100000, true);
+    timerAlarmEnable(timer);
+}
+
+void IRAM_ATTR Geneva::pulseEncoderA()
+{
+    count_a = count_a + dir;
 }
 
-float Geneva::getCurrentAngle()
+void IRAM_ATTR Geneva::pulseEncoderB()
 {
-    // TODO(#2458): convert from EncoderValue to angle
-    return 0;
+    count_b = count_b + dir;
 }
 
-void Geneva::performWhenDone(void (*isr)(void))
+void IRAM_ATTR Geneva::onTimer()
 {
-    timerRestart(timer);
+    // stop motor when it stalls aka in side slots
+    // or if its centering it should be a fixed amount
+    if (((currentSlot == TbotsProto_Geneva_Slot_LEFT ||
+          currentSlot == TbotsProto_Geneva_Slot_RIGHT) &&
+         ((count_a != 0 && prev_count_a == count_a) ||
+          (count_b != 0 && prev_count_b == count_b))) ||
+        (currentSlot == TbotsProto_Geneva_Slot_CENTRE &&
+         ((dir == 1 && count_a >= CENTERING_VALUE_FROM_LEFT) ||
+          (dir == -1 && count_a <= CENTERING_VALUE_FROM_RIGHT))))
+    {
+        digitalWrite(PWM, LOW);
+        count_a = 0;
+        count_b = 0;
+        if (rotation_done_callback)
+        {
+            rotation_done_callback();
+            rotation_done_callback = NULL;
+        }
+    }
 
-    timerAttachInterrupt(timer, isr, true);
-    // TODO(#2458): calculate delay based on start position and end position
-    // TODO(#2458): take max of time to setChargeDoneCallbackOnce once known
-    timerAlarmWrite(timer, 1000000 /* 1s */, false);
-    timerAlarmEnable(timer);
+    prev_count_a = count_a;
+    prev_count_b = count_b;
+}
+
+TbotsProto_Geneva_Slot Geneva::getCurrentSlot()
+{
+    return currentSlot;
+}
+void Geneva::setRotationDoneCallbackOnce(void (*rotation_done_callback)())
+{
+    Geneva::rotation_done_callback = rotation_done_callback;
+}
+
+void Geneva::rotateLeft()
+{
+    dir = 1;
+    digitalWrite(DIR, LOW);
+    digitalWrite(PWM, HIGH);
+}
+
+void Geneva::rotateRight()
+{
+    dir = -1;
+    digitalWrite(DIR, HIGH);
+    digitalWrite(PWM, HIGH);
 }
 
-void Geneva::setAngle(float angle_deg, void (*isr)(void))
+void Geneva::setSlot(TbotsProto_Geneva_Slot slot)
 {
-    // TODO(#2458): Set angle through encoderValue
-    // TODO(#2458): Set up values for isr (can't add args but can access static members)
-    performWhenDone(isr);
+    switch (slot)
+    {
+        case TbotsProto_Geneva_Slot_LEFT:
```

Comment:
rather than doing relative movements, I think we can "home" the geneva. At the very beginning, we can move the geneva to one extreme and zero out the encoder values. And then from there, each slot will be a fixed amount of encoder clicks which can keep track of.

If its different from robot to robot, we can make use of redis (but I don't think it will be)

Reply:
We wouldn't zero out the encoder when we switch slots. We just go 

~~~
if target < current: go left;
if target > current go_right; 
if target ~ current: stop
~~~

And each slot would just have a lookup for how many clicks it takes to get there (which we can calibrate)